Serialize Message Metadata into Flatbuffers

Benjamin-Philip opened the pull request:
This commit adds support to serialize message metadata into flatbuffers.

Benjamin-Philip committed:
Add Erlang side testing of serialize_message

Benjamin-Philip committed:
Fix Clippy errors

Benjamin-Philip commented:
@josevalim This PR is still a draft...

josevalim commented:
And the plan is looking good so far!

Benjamin-Philip commented:
I don't think I'm going make any Erlang changes after this, so just as well I guess. :)

Benjamin-Philip committed:
Serialize Record batches

This commit adds the implementation and tests to serialize a RecordBatch.

Benjamin-Philip committed:
Serialize Types

Benjamin-Philip committed:
Serialize Schema Fields

In addition to serializing schema fields, there have been some soft
changes. In creating this commit, I've come to realise that Atoms are a
bad thing to keep in Rust structs. So, in this commit, I've also changed
Name and Dictionary to use enums to represent undefined instead of an
atom.

Benjamin-Philip committed:
Serialize Schemas

Benjamin-Philip committed:
Reduce test data duplication

Benjamin-Philip committed:
Fix failing test

Benjamin-Philip committed:
Serialize Headers

Benjamin-Philip committed:
Serialize Messages to their Rust form

This commit serializes messages to their planus, or Rust form.
It does not serialize it to proper flatbuffers yet.

Benjamin-Philip committed:
Serialize messages to IPC

This commit adds the final commit required to serialize message metadata
to IPC. Note though, that this function has not been integrated with
`serde_arrow_ipc_message:to_ipc` function.

Benjamin-Philip committed:
Fix lints

Benjamin-Philip commented:
@josevalim, There haven't been any Erlang changes since you last reviewed, but you still may want to take a look.

polvalente commented on native/arrow_format_nif/src/utils.rs:
```diff
@@ -1,3 +1,292 @@
 pub type CustomMetadata = Vec<HashMap<String, String>>;
+
+/// Returns an example `Message` of a `Schema`
+pub fn schema() -> Message {
```
I generally don't like "utils" as a module name because it tends to become a blackhole of all things that don't quite have a place.

If I understood correctly, `schema` and `message_batch` a just dummy functions that will be replaced where they are currently called by actual code in the future.

If this is indeed the case, I think it would be best if they were prefixed with `dummy_` and set in the `message` module instead.

Benjamin-Philip replied:
They are functions which return testing data, and are used solely in testing.
Presently they are are used in the `test_decode` function in `crate`, and in various Rust specific tests.

Should I still prefix them with `dummy` and move them or is keeping them in `utils` or in a `test_utils` good enough?

polvalente replied:
Should `test_decode` actually be exposed in the NIF?

Benjamin-Philip replied:
Yes, it's a function to test if everything decodes properly into Erlang.

This PR only implements the encoding part of this NIF, and I will have to deal with decoding later.

polvalente replied:
So that function will be removed as soon as the decoding is implemented, right?
I believe it's fine to leave it where it is then :)

Benjamin-Philip replied:
I've put the `arrow_*` into a `test::fixtures` module in `lib.rs`. I hoping this is what you had in mind when you said `test::fixtures`? I wasn't able to find a way to keep them in a file in the `tests` dir and use them in unit tests in the `src` dir.

Benjamin-Philip committed:
Move utils::arrow_* functions to test::fixtures

Since these are test only functions, it's better to keep them in a test
specific module.

Benjamin-Philip committed:
Remove FIXMEs for dead code

Since the code in schema and record_batch are no longer dead, we need
not allow dead code in the clippy checks.

Benjamin-Philip committed:
cargo fmt

polvalente commented on native/arrow_format_nif/src/lib.rs:
```diff
@@ -49,10 +46,156 @@ mod atoms {
 }
 
+#[cfg(test)]
+pub mod test {
```
I believe you could have this extracted to a `test.rs` file in the `src` directory. This is a non-blocker, though :)
